blog/2024-04-modern-angular-bm/README.md

```diff
@@ -0,0 +1,938 @@
+import { /* ... */, effect, inject } from '@angular/core';
+// ...
+
+@Directive({ /* ... */ })
+export class LoggedinOnlyDirective {
+  // ...
+  constructor() {
+    effect(() => {
+      if (this.authService.isAuthenticated()) {
+        this.viewContainer.createEmbeddedView(this.template);
+      } else {
+        this.viewContainer.clear();
+      }
+    });
+  }
+}
+```
+
+### Signal Inputs
+
+Mit dem Minor-Release von Angular 17.2.0 wurde eine Alternative zum bisherigen `@Input()`-Dekorator auf Basis von Signals eingeführt, siehe die [offizielle Information im Angular-Blog](https://blog.angular.io/signal-inputs-available-in-developer-preview-6a7ff1941823).
+Der übergebene Wert eines Komponenten-Inputs wird damit direkt als Signal erfasst:
+
+```ts
+isbn = input() // InputSignal<unknown>
+isbn = input<string>() // InputSignal<string | undefined>
+isbn = input.required() // InputSignal<unknown>
+isbn = input.required<string>() // InputSignal<string>
+isbn = input('3864909465') // InputSignal<string>
+```
+
+Wir können einige Stellen unserer Anwendung auf die neuen *Signal-based Inputs* umbauen.
+Als erstes Beispiel wollen wir die `FormErrorsComponent` nutzen.
+Wir nutzen hier jeweils `input.required()`, da beide Informationen zwingend bei Verwendung der Komponente gesetzt werden müssen.
+Um das Signal zu lesen, dürfen wir im Getter `errors` die Funktionsklammern nicht vergessen.
+
+```ts
+import { Component, inject, input } from '@angular/core';
+// ...
+
+@Component({ /* ... */ })
+export class FormErrorsComponent {
+  controlName = input.required<string>();
+  messages = input.required<{ [errorCode: string]: string }>();
+  // ...
+
+  get errors(): string[] {
+    const control = this.form.control.get(this.controlName());
+    // ...
+    return Object.keys(control.errors).map(errorCode => {
+      return this.messages()[errorCode];
+    });
+  }
+}
+```
+
+Nicht immer wollen wir, dass der Name des Propertys in der Komponente dem Namen entspricht, den wir beim Property Binding von außen verwenden.
+Diesen Fall haben wir bereits in der `ConfirmDirective` berücksichtigt.
+Auch mit Signal Inputs lässt sich ein `alias` konfigurieren.
+
+Wenn wir die `ConfirmDirective` verwenden, wollen wir den Text an das Property `bmConfirm` übergeben, das dem Attributnamen der Direktive entspricht.
+Innerhalb der Direktivenklasse soll dieses Property jedoch `confirmText` lauten.
+Ein Alias ermöglicht uns diese Verbindung:
+
+```ts
+import { /* ... */, input } from '@angular/core';
+
+@Directive({
+  selector: '[bmConfirm]',
+  standalone: true
+})
+export class ConfirmDirective {
+  confirmText = input.required<string>({ alias: 'bmConfirm' });
+  // ...
+  @HostListener('click') onClick() {
+    if (window.confirm(this.confirmText())) {
+      this.confirm.emit();
+    }
+  }
+}
+```
+
+> Das Projekt _ngxtension_ stellt zur Migration auf Signal Inputs auch ein [Schematic](https://ngxtension.netlify.app/utilities/migrations/signal-inputs-migration/) bereit.
```

Comment:
Sollten wir hier noch konkreter werden?

```bash
npm i -D ngxtension
npx ng g ngxtension:convert-signal-inputs
```

Reply:
Weiß ich nicht so richtig … ich kann mir vorstellen, dass wir unten einen weiteren Abschnitt machen, in dem wir alle diese Befehle auflisten:

> Migrations-Schematics von ngxtension verwenden

etwa:
In den vorherigen Abschnitten haben wir erwähnt, dass das Paket [ngxtension](LINK) Schematics für die Migration auf einige neuere Angular-Features anbietet. Es handelt sich nicht um ein offizielles Projekt von Angular, sondern stammt aus der Community.

Wir müssen das Paket zunächst installieren:

```bash
npm i -D ngxtension
```

Anschließend können wir die Migration mithilfe von `ng generate` anstoßen:

```bash
# von @Input zu Signal Inputs
ng g ngxtension:convert-signal-inputs

# von @Output zu Functional Outputs
ng g ngxtension:convert-outputs

# von  Constructor Injection zu inject()
ng g ngxtension:convert-di-to-inject
```

Reply:
und oben jeweils kurz drauf verweisen

Reply:
Das finde ich für den Lesefluss besser, als die Befehle direkt zu erwähnen